### ISSUE TITLE
LCC (LGPLv3+) includes headers from Triangulation_3 (GPLv3+)

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#1492]
The file `<CGAL/Linear_cell_complex_constructors.h>` from LCC (under LGPLv3+) includes:
```C++
#include <CGAL/Triangulation_2.h>
#include <CGAL/Triangulation_3.h>
```
from GPLv3+ packages. That breaks the license, making the file GPLv3. That also breaks a license checking feature we are trying to integrate into CGAL.

That header is documented in [Constructions for Linear Cell Complex](http://doc.cgal.org/latest/Linear_cell_complex/group__PkgLinearCellComplexConstructions.html).

**Comment by gdamiand** [comment]
Thanks @lrineau for pointing that error, sorry for this mistake.
Note that it is also the same problem with Polyhedron_3.h.

First this concerns only Linear_cell_complex_constructors.h and not Linear_cell_complex.h itself which is nice.

I think the simplest solution is to move the file Linear_cell_complex_constructors.h from CGAL/include to CGAL/example.

Of course this could break backward compatibility for users, but in this case it is easy to correct the error by just copying the Linear_cell_complex_constructors.h.

@lrineau What do you think ?

**Comment by afabri** [comment]
We had the same problem with `boost::graph_traits<CGAL:Polyhedron_3>` which was in the BGL package. The solution was to move the file from the BGL package to the Polyhedron package.  It then stays in `CGAL/include/CGAL/` that is there is not even a backward compatibility issue.

**Comment by gdamiand** [comment]
In my case, the file must be split in two parts : one part will go into T3 package while the second one will go into Polyhedron package.

**Comment by afabri** [comment]
If it is just file editing and moving, we can do that for you.

**Comment by gdamiand** [comment]
Ok nice, perfect. Thanks.

**Comment by afabri** [comment]
In fact, as soon as we split the file we have to document that users have to `#include` other files, that is we break backward compatibility a little bit.

**Comment by mglisse** [comment]
> That breaks the license, making the file GPLv3

That's an oversimplification. It is legal to have an LGPL file include a GPL file. The result is GPL, that's fine as long as the doc avoids misleading users. If a user explicitly builds a T3 and passes it to LCC, they are also directly using the T3 package, so they know they are using GPL code.
(you can still split/move the file, though putting it in the T3 package would likely require that LIRIS assigns the copyright to Inria, which sounds like a lot of trouble)

**Comment by lrineau** [comment]
Of course you are right about me simplifying. It was on purpose.

And you are also right about the split/move of the file, that would require paperwork. We'll have to come up with a solution, if possible.

**Comment by gdamiand** [comment]
What about my solution ? (move this file in example). I am not sure many users use these functions...

**Comment by mglisse** [comment]
@gdamiand It is a bit sad to remove an interface for this kind of reason. If you think they are really useless, sure, move them to examples. But otherwise, I'd rather we find a general solution to this problem. If we have a policy that a package can only have 1 license and that an LPGL package cannot use a GPL package (even in an optional way?), we could always have 2 packages, LCC and LCC_GPL, where the second one contains just the problematic file, with an artificial GPL license to please the license checking script... :disappointed:
In any case, it would make sense to split the _constructors.h file to avoid including the world in there, unless some light forward declarations could replace the heavy includes.

**Comment by lrineau** [comment]
I agree with Marc that it would be a lose to remove a useful header from the distribution of CGAL headers.

From the discussion, we can see that, even if the header is a member of LCC, it is more a way for triangulation users to use LCC, and does not mean at all that T_3 is a dependency of LCC. Let me remove that header from the scope of the dependency checking, and let's keep it where it is. Even if the header if LGPL, its dependency to a headers makes it practically GPL, but as Marc pointed out that is not a real issue.

Marc, Guillaume, Andreas, do you all agree that we can close this issue as not-a-bug?

**Comment by afabri** [comment]
The split in several files  import_from_XX.h makes a lot of sense as it makes it much more modular, and as then the file name corresponds to the function name.

@gdamiand  If you want GeometryFactory rewrites these functions from scratch (and as a bonus we make them faster switching to hash maps) , so that there is no liris copyright header in the file.  There is no money to earn from these files, as they are not core CC or LCC  functionality, but glue.

**Comment by mglisse** [comment]
> Marc, Guillaume, Andreas, do you all agree that we can close this issue as not-a-bug?
If splitting the file then marking the 2 that depend on Triangulation_3 or Polyhedron as exceptions in your scripts works for you, that seems good to me: the licensing situation is a bit clearer, without any awkward consequences. :+1:

**Comment by gdamiand** [comment]
I agree that we can close this issue.

**Comment by lrineau** [comment]
Fixed by cf8c1a6 from PR #1530.
